kubetest2/pkg/exec/exec.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"os"
+)
+
+// Cmd abstracts over running a command somewhere, this is useful for testing
+type Cmd interface {
+	Run() error
+	// Each entry should be of the form "key=value"
+	SetEnv(...string) Cmd
+	SetStdin(io.Reader) Cmd
+	SetStdout(io.Writer) Cmd
+	SetStderr(io.Writer) Cmd
+}
+
+// Cmder abstracts over creating commands
+type Cmder interface {
+	// command, args..., just like os/exec.Cmd
+	Command(string, ...string) Cmd
+}
+
+// DefaultCmder is a LocalCmder instance used for convienience, packages
+// originally using os/exec.Command can instead use pkg/kind/exec.Command
+// which forwards to this instance
+// TODO(bentheelder): swap this for testing
+// TODO(bentheelder): consider not using a global for this :^)
+var DefaultCmder = &LocalCmder{}
+
+// Command is a convience wrapper over DefaultCmder.Command
+func Command(command string, args ...string) Cmd {
+	return DefaultCmder.Command(command, args...)
+}
+
+// CombinedOutputLines is like os/exec's cmd.CombinedOutput(),
+// but over our Cmd interface, and instead of returning the byte buffer of
+// stderr + stdout, it scans these for lines and returns a slice of output lines
+func CombinedOutputLines(cmd Cmd) (lines []string, err error) {
```

Comment:
it would be nicer to be able to create a Cmd instance and call: `myCmd.CombinedOutputLines()`
applies to InheritOutput too.

this would mean moving these functions as part of the interface and implementing them for LocalCmd?

Reply:
> this would mean moving these functions as part of the interface and implementing them for LocalCmd?

these functions are specifically _NOT_ in the interface because having a smaller interface makes it easier to implement widely. this function only needs the smaller interface to be implemented in order to work.

Reply:
`myCmd.CombinedOutputLines()` is really similar to `exec.CombinedOutputLines(myCmd)` but the latter reduces complexity to implement and keeps the interface focused. we can implement similar ones like `exec.StdoutLines(myCmd)` easily over the same interface.